### PR TITLE
Adapter for Cottontail-DB

### PR DESCRIPTION
**Pull request by schoenja**
This PR adds an adapter for Cottontail-DB, a columnar data store optimized for multimedia-retrieval.

### CODE REVIEW

File: cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/rel/CottontailValues.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+
+        public Translator( RelDataType rowType ) {
+            this.rowType = rowType;
+            List<Pair<String, String>> pairs = Pair.zip( rowType.getFieldList().stream().map( RelDataTypeField::getPhysicalName ).collect( Collectors.toList() ), rowType.getFieldNames() );
+            this.fieldNames = pairs.stream().map( it -> it.left != null ? it.left : it.right ).collect( Collectors.toList() );
+        }
+
+
+    }
```
Comment (vogti) [inline cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/rel/CottontailValues.java:125]: Is this code still used?

**Review by vogti — CHANGES_REQUESTED** [review]
Thx @schoenja for this PR. Could you please have a look at my questions? Furthermore, it would be very nice if you would find the time to scroll through your changes and remove outdated todos and remove out-commented code which is of no further use (please do not remove code which might be useful in future).

File: cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/rules/CottontailRules.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+
+    public static List<RelOptRule> rules( CottontailConvention out, RelBuilderFactory relBuilderFactory ) {
+        return ImmutableList.of(
+                new CottontailToEnumerableConverterRule( out, relBuilderFactory ),
+                new CottontailValuesRule( out, relBuilderFactory ),
+                new CottontailTableModificationRule( out, relBuilderFactory ),
+                new CottontailProjectRule( out, relBuilderFactory ),
+//                new CottontailSortRule( out, relBuilderFactory ),
```
[the remark this answers is not in the document]
Comment (schoenja) [reply cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/rules/CottontailRules.java:42]: Pure sorting is not supported by Cottontail-DB. The rule has been replaced by the `CottontailSortAndProjectRule`.

File: cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/CottontailWrapper.java [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        try {
+            stub.createIndex( createMessage );
+        } catch ( StatusRuntimeException e ) {
+            if ( e.getStatus().getCode() == Status.ALREADY_EXISTS.getCode() ) {
+                log.warn( "Index on {}.{} was not created because it already exists", createMessage.getIndex().getEntity().getName(), createMessage.getColumnsList().toString() );
+                return;
+            }
+            log.error( "Caught exception", e );
+        }
+    }
```
[the remark this answers is not in the document]
Comment (schoenja) [reply cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/CottontailWrapper.java:110]: This method originates from the original Wrapper in the Cineast project. I currently can't say whether Cottontail-DB indices and Polypheny-DB indices are compatible.

File: cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/rel/CottontailValues.java (hunk 1 above)
Comment (schoenja) [reply cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/rel/CottontailValues.java:125]: Nope, will be deleted.

File: cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/rel/CottontailLimit.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+
+    @Override
+    public void implement( CottontailImplementContext context ) {
+        context.visitChild( 0, getInput() );
+        if ( offset != null ) {
+//            context.offset = RexLiteral.intValue( offset );
+        }
+        if ( fetch != null ) {
+//            context.limit = RexLiteral.intValue( fetch );
```
[the remark this answers is not in the document]
Comment (schoenja) [reply cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/rel/CottontailLimit.java:66]: No, the entire `CottontailLimit` class is no longer required, as those limits are combined with `CottontailSortAndProject`.

File: cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/CottontailStore.java [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+
+        queryResponse.forEachRemaining( queryResponseMessage -> {
+            for ( Tuple tuple : queryResponseMessage.getResultsList() ) {
+                Map<String, Data> dataMap = new HashMap<>( tuple.getDataMap() );
+                Object value = dataMap.get( newPhysicalColumnName );
+//                dataMap.put( newPhysicalColumnName, CottontailTypeConversionUtil.convertValue( value,  ) );
+
+                inserts.add( InsertMessage.newBuilder().setTuple( Tuple.newBuilder().putAllData( dataMap ).build() ).setFrom( from ).build() );
+            }
+        } );
```
[the remark this answers is not in the document]
Comment (schoenja) [reply cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/CottontailStore.java:510]: Yes, it is currently unfinished. Do you want me to just delete it or keep it?

[the remark this answers is not in the document]
Comment (vogti) [reply cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/CottontailStore.java:510]: Please add a comment on top of the method and keep the code.

File: cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/CottontailWrapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vogti) [reply cottontail-adapter/src/main/java/org/polypheny/db/adapter/cottontail/CottontailWrapper.java:110]: The implementation for managing store level indexes is very generic. There should be no compatibility issues. Adding support for store level indexes only requires implementing the methods in the store interface. I can do that, but I need a _dropIndex_ method and a short example on how to use these methods.

**Review by vogti — APPROVED** [review]
Thx @schoenja for this PR!

**Pull request merged**